File: packages/flutter_shared/lib/tools/timestamp_tool.dart

/// 时间戳工具 — 秒级/毫秒级互转
class TimestampTool {
  TimestampTool._();

  /// 时间戳 → DateTime（自动识别秒/毫秒）
  static ({DateTime dateTime, bool isMs}) toDateTime(num timestamp) {
    if (timestamp > 1e12) {
      // 毫秒级 (13 digits)
      return (
        dateTime: DateTime.fromMillisecondsSinceEpoch(timestamp.toInt()),
        isMs: true,
      );
    } else {
      // 秒级 (10 digits)
      return (
        dateTime: DateTime.fromMillisecondsSinceEpoch(timestamp.toInt() * 1000),
        isMs: false,
      );
    }
  }

  /// DateTime → 秒级时间戳
  static int toSeconds(DateTime dt) => dt.millisecondsSinceEpoch ~/ 1000;

  /// DateTime → 毫秒级时间戳
  static int toMilliseconds(DateTime dt) => dt.millisecondsSinceEpoch;

  /// 格式化为 ISO 8601
  static String toIso(DateTime dt) => dt.toIso8601String();

  /// 格式化为中国常用格式: 2024-01-15 10:30:00
  static String toChinese(DateTime dt) {
    return '${dt.year.toString().padLeft(4, '0')}-'
        '${dt.month.toString().padLeft(2, '0')}-'
        '${dt.day.toString().padLeft(2, '0')} '
        '${dt.hour.toString().padLeft(2, '0')}:'
        '${dt.minute.toString().padLeft(2, '0')}:'
        '${dt.second.toString().padLeft(2, '0')}';
  }

  /// 格式化为美国格式: 01/15/2024 10:30:00 AM
  static String toAmerican(DateTime dt) {
    final hour = dt.hour > 12 ? dt.hour - 12 : (dt.hour == 0 ? 12 : dt.hour);
    final ampm = dt.hour >= 12 ? 'PM' : 'AM';
    return '${dt.month.toString().padLeft(2, '0')}/'
        '${dt.day.toString().padLeft(2, '0')}/'
        '${dt.year} '
        '$hour:'
        '${dt.minute.toString().padLeft(2, '0')}:'
        '${dt.second.toString().padLeft(2, '0')} $ampm';
  }

  /// 相对时间描述
  static String relative(DateTime dt) {
    final now = DateTime.now();
    final diff = dt.difference(now);
    final seconds = diff.inSeconds.abs();

    if (seconds < 60) return '几秒${diff.isNegative ? '前' : '后'}';
    if (seconds < 3600) return '${seconds ~/ 60} 分钟${diff.isNegative ? '前' : '后'}';
    if (seconds < 86400) return '${seconds ~/ 3600} 小时${diff.isNegative ? '前' : '后'}';
    if (seconds < 172800) return diff.isNegative ? '昨天' : '明天';
    if (seconds < 604800) return '${seconds ~/ 86400} 天${diff.isNegative ? '前' : '后'}';
    if (seconds < 2592000) return '${seconds ~/ 604800} 周${diff.isNegative ? '前' : '后'}';
    if (seconds < 31536000) return '${seconds ~/ 2592000} 个月${diff.isNegative ? '前' : '后'}';
    return '${seconds ~/ 31536000} 年${diff.isNegative ? '前' : '后'}';
  }

  /// 获取当前时间戳（秒）
  static int nowSeconds() => DateTime.now().millisecondsSinceEpoch ~/ 1000;

  /// 获取当前时间戳（毫秒）
  static int nowMilliseconds() => DateTime.now().millisecondsSinceEpoch;
}

File: miniapp/src/utils/tools/timestampTool.ts
/**
 * 时间戳工具 — 秒级/毫秒级互转
 * 移植自 flutter_shared/tools/timestamp_tool.dart
 */

export interface TimestampDateTime {
  dateTime: Date;
  isMs: boolean;
}

export const TimestampTool = {
  /** 时间戳 → Date（自动识别秒/毫秒） */
  toDateTime(timestamp: number): TimestampDateTime {
    if (timestamp > 1e12) {
      return { dateTime: new Date(timestamp), isMs: true };
    }
    return { dateTime: new Date(timestamp * 1000), isMs: false };
  },

  /** Date → 秒级时间戳 */
  toSeconds(dt: Date): number {
    return Math.floor(dt.getTime() / 1000);
  },

  /** Date → 毫秒级时间戳 */
  toMilliseconds(dt: Date): number {
    return dt.getTime();
  },

  /** ISO 8601 */
  toIso(dt: Date): string {
    return dt.toISOString();
  },

  /** 中国常用格式: 2024-01-15 10:30:00 */
  toChinese(dt: Date): string {
    const pad = (n: number) => n.toString().padStart(2, '0');
    return `${dt.getFullYear()}-${pad(dt.getMonth() + 1)}-${pad(dt.getDate())} ${pad(dt.getHours())}:${pad(dt.getMinutes())}:${pad(dt.getSeconds())}`;
  },

  /** 美国格式: 01/15/2024 10:30:00 AM */
  toAmerican(dt: Date): string {
    const pad = (n: number) => n.toString().padStart(2, '0');
    const hour = dt.getHours() > 12 ? dt.getHours() - 12 : (dt.getHours() === 0 ? 12 : dt.getHours());
    const ampm = dt.getHours() >= 12 ? 'PM' : 'AM';
    return `${pad(dt.getMonth() + 1)}/${pad(dt.getDate())}/${dt.getFullYear()} ${hour}:${pad(dt.getMinutes())}:${pad(dt.getSeconds())} ${ampm}`;
  },

  /** 相对时间描述 */
  relative(dt: Date): string {
    const now = new Date();
    const diff = dt.getTime() - now.getTime();
    const seconds = Math.abs(Math.floor(diff / 1000));
    const isPast = diff < 0;

    if (seconds < 60) return `几秒${isPast ? '前' : '后'}`;
    if (seconds < 3600) return `${Math.floor(seconds / 60)} 分钟${isPast ? '前' : '后'}`;
    if (seconds < 86400) return `${Math.floor(seconds / 3600)} 小时${isPast ? '前' : '后'}`;
    if (seconds < 172800) return isPast ? '昨天' : '明天';
    if (seconds < 604800) return `${Math.floor(seconds / 86400)} 天${isPast ? '前' : '后'}`;
    if (seconds < 2592000) return `${Math.floor(seconds / 604800)} 周${isPast ? '前' : '后'}`;
    if (seconds < 31536000) return `${Math.floor(seconds / 2592000)} 个月${isPast ? '前' : '后'}`;
    return `${Math.floor(seconds / 31536000)} 年${isPast ? '前' : '后'}`;
  },

  nowSeconds(): number {
    return Math.floor(Date.now() / 1000);
  },

  nowMilliseconds(): number {
    return Date.now();
  },
};
